### PR TITLE
Support custom SendDecorator chains via context

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Added `autorest.WithSendDecorators` and `autorest.GetSendDecorators` for
adding and retrieving a custom chain of SendDecorators to the provided
context.
Added `autorest.DoRetryForStatusCodesWithCap` and
`autorest.DelayForBackoffWithCap` to enforce an upper bound on the
duration between retries.
Fixed up some code comments.

Thank you for your contribution to Go-AutoRest! We will triage and review it as soon as we can.

As part of submitting, please make sure you can make the following assertions:
 - [ ] I've tested my changes, adding unit tests if applicable.
 - [ ] I've added Apache 2.0 Headers to the top of any new source files.
 - [ ] I'm submitting this PR to the `dev` branch, except in the case of urgent bug fixes warranting their own release.
 - [ ] If I'm targeting `master`, I've updated [CHANGELOG.md](https://github.com/Azure/go-autorest/blob/master/CHANGELOG.md) to address the changes I'm making.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/go-autorest/issues/398 and https://github.com/Azure/go-autorest/issues/409

File: autorest/sender.go [hunk 1]
```diff
@@ -211,43 +229,59 @@ func DoRetryForAttempts(attempts int, backoff time.Duration) SendDecorator {
...
-			return resp, err
+		return SenderFunc(func(r *http.Request) (*http.Response, error) {
+			return doRetryForStatusCodesImpl(s, r, false, attempts, backoff, 0, codes...)
 		})
 	}
 }
 
+// DoRetryForStatusCodesWithCap returns a SendDecorator that retries for specified statusCodes for up to the
+// specified number of attempts, exponentially backing off between requests using the supplied backoff
+// time.Duration (which may be zero). To cap the maximum possible delay specify a value greater than zero for
```
Comment (vladbarosan) [inline autorest/sender.go]: this is the total maximum duration for this right ? maybe we can clarify it a little

File: autorest/sender.go [hunk 2]
```diff
@@ -211,43 +229,59 @@ func DoRetryForAttempts(attempts int, backoff time.Duration) SendDecorator {
...
+// cap. Retrying may be canceled by cancelling the context on the http.Request.
+func DoRetryForStatusCodesWithCap(attempts int, backoff, cap time.Duration, codes ...int) SendDecorator {
+	return func(s Sender) Sender {
+		return SenderFunc(func(r *http.Request) (*http.Response, error) {
+			return doRetryForStatusCodesImpl(s, r, true, attempts, backoff, cap, codes...)
+		})
+	}
+}
+
+func doRetryForStatusCodesImpl(s Sender, r *http.Request, countAll bool, attempts int, backoff, cap time.Duration, codes ...int) (resp *http.Response, err error) {
```
Comment (vladbarosan) [inline autorest/sender.go]: does it make sense to pull these in an options type ? seem a little too many params

Comment (vladbarosan) [reply autorest/sender.go]: and countAll, maybe count429Codes, or countAllCodes ?

File: autorest/sender.go [hunk 3]
```diff
@@ -325,8 +359,22 @@ func WithLogging(logger *log.Logger) SendDecorator {
...
+	return DelayForBackoffWithCap(backoff, 0, attempt, cancel)
+}
+
+// DelayForBackoffWithCap invokes time.After for the supplied backoff duration raised to the power of
+// passed attempt (i.e., an exponential backoff delay). Backoff duration is in seconds and can set
+// to zero for no delay. To cap the maximum possible delay specify a value greater than zero for cap.
+// The delay may be canceled by closing the passed channel. If terminated early, returns false.
+// Note: Passing attempt 1 will result in doubling "backoff" duration. Treat this as a zero-based attempt
+// count.
+func DelayForBackoffWithCap(backoff, cap time.Duration, attempt int, cancel <-chan struct{}) bool {
```
Comment (vladbarosan) [inline autorest/sender.go]: i was thinking kind of the pattern with `COntextWithTimeout`, we can name it the same ?

Comment (vladbarosan) [reply autorest/sender.go]: i.e. `WithTimeout`

File: autorest/sender.go (hunk 1 above)
Comment (jhendrixMSFT) [reply autorest/sender.go]: Not the aggregate total, the total between iterations.

File: autorest/sender.go (hunk 3 above)
Comment (jhendrixMSFT) [reply autorest/sender.go]: The `WithCap` is the suffix to differentiate it from `DelayForBackoff`.  Or maybe I misunderstand?

File: autorest/sender.go (hunk 1 above)
Comment (jhendrixMSFT) [reply autorest/sender.go]: Will update the comment with this info.

**Pull request merged**